auacm/angular/src/app/competition.service.ts

```diff
@@ -0,0 +1,341 @@
+import { Injectable } from '@angular/core';
+import { Http, Request, Response, Headers, URLSearchParams } from '@angular/http';
+import { Subject } from 'rxjs';
+
+import { WebsocketService } from './websocket.service'; 
+
+import { CompetitionProblem, Competition, CompetitionTeam, TeamProblemData } from './models/competition';
+import { RecentSubmission } from './models/submission';
+import { Problem } from './models/problem';
+import { SimpleUser, WebsocketRegisteredUser } from './models/user';
+
+
+/**
+ * This class manages the connection to the competition part of the backend.
+ * You can add, edit, or remove competitions, update teams, and handle websocket
+ * messages about competitions.
+ * 
+ * Websocket messages:
+ *   - system_time
+ *   - status
+ *   - new_user
+ */
+@Injectable()
+export class CompetitionService {
+
+  // Websocket subject for client time offset
+  private clientTimeOffset: number = 0;
+  timeOffsetSource: Subject<number> = new Subject<number>();
+
+  // Websocket subject for competition data
+  private competition: Competition = new Competition();
+  competitionSource: Subject<Competition> = new Subject<Competition>();
+
+  // Websocket subject for competition teams
+  competitionTeamSource: Subject<WebsocketRegisteredUser> = new Subject<WebsocketRegisteredUser>();
+
+  /**
+   * The constructor for CompetitionService. This handles initializing the
+   * connection with the websocket endpoint and handling the responses from
+   * the websocket. It currently handles the system_time, status, new_user
+   * websocket messages.
+   */
+  constructor(private _http: Http, private _websocketService: WebsocketService) {
+    this._websocketService.connect(window.location.host + '/websocket').subscribe(data => {
+        // update the client's time offset and scoreboard data
+        var response = JSON.parse(data.data);
+        var responseData = response.data;
+        var viewed = [];
```

Comment:
It seems to me that viewed will contain all submission Ids that occurred during the contest. In the event of a large contest, with many submissions, the viewed.indexOf, which is O(N), could start to slow down. Do you think changing this to a set would be a worthwhile modification?